File: app/src/main/java/com/hanix/randomgame/common/utils/ToastUtil.java
package com.hanix.randomgame.common.utils;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import com.hanix.randomgame.R;
import com.hanix.randomgame.common.app.GLog;

import static android.view.Gravity.BOTTOM;


public class ToastUtil {

    private static Toast mToast;

    /**
     * LENGTH_SHORT Toast
     *
     * @param context
     * @param msg
     */
    public static void showToastS(final Context context, final String msg) {
        new Handler(Looper.getMainLooper()).post(() -> {
            hideToast();
            mToast = Toast.makeText(context, msg, Toast.LENGTH_SHORT);
            mToast.show();
        });
    }

    /**
     * LENGTH_LONG Toast
     *
     * @param context
     * @param msg
     */
    public static void showToastL(final Context context, final String msg) {
        new Handler(Looper.getMainLooper()).post(() -> {
            hideToast();
            mToast = Toast.makeText(context, msg, Toast.LENGTH_LONG);
            mToast.show();
        });
    }

    /**
     * 토스트 전부 내리기
     */
    public static void hideToast() {
        if (mToast != null) {
            GLog.i("Toast hide true!");
            mToast.cancel();
        }
    }

    /**
     * @param context
     * @param msg
     * @param isHtml
     */
    public static void showToastPresent(final Context context, final String msg, final boolean isHtml) {
        new Handler(Looper.getMainLooper()).post(() -> {
            if (isHtml) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    Toast.makeText(context, Html.fromHtml(msg, Html.FROM_HTML_MODE_LEGACY), Toast.LENGTH_LONG).show();
                }
            } else {
                Toast.makeText(context, msg, Toast.LENGTH_LONG).show();
            }
        });
    }

    public static void setToastLayout(Activity activity, String msg, Context context) {
        LayoutInflater inflater = activity.getLayoutInflater();
        View layout = inflater.inflate(R.layout.toast_layout, activity.findViewById(R.id.toastLayout));

        TextView toastTv = layout.findViewById(R.id.toastTv);
        toastTv.setText(msg);

        final Toast toast = Toast.makeText(context, msg, Toast.LENGTH_SHORT);
        toast.setGravity(BOTTOM, 0, 100);
        toast.setDuration(Toast.LENGTH_SHORT);
        toast.setView(layout);
        toast.show();
    }
}

File: app/src/main/java/com/hanix/randomgame/common/utils/ToastUtil.kt
package com.hanix.randomgame.common.utils

import android.app.Activity
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.text.Html
import android.view.Gravity
import android.widget.Toast
import com.hanix.randomgame.R
import com.hanix.randomgame.common.app.GLog.i

object ToastUtil {
    private var mToast: Toast? = null

    /**
     * LENGTH_SHORT Toast
     *
     * @param context
     * @param msg
     */
    fun showToastS(context: Context?, msg: String?) {
        Handler(Looper.getMainLooper()).post {
            hideToast()
            mToast = Toast.makeText(context, msg, Toast.LENGTH_SHORT)
            mToast?.show()
        }
    }

    /**
     * LENGTH_LONG Toast
     *
     * @param context
     * @param msg
     */
    fun showToastL(context: Context?, msg: String?) {
        Handler(Looper.getMainLooper()).post {
            hideToast()
            mToast = Toast.makeText(context, msg, Toast.LENGTH_LONG)
            mToast?.show()
        }
    }

    /**
     * 토스트 전부 내리기
     */
    private fun hideToast() {
        if (mToast != null) {
            i("Toast hide true!")
            mToast!!.cancel()
        }
    }

    /**
     * @param context
     * @param msg
     * @param isHtml
     */
    fun showToastPresent(context: Context?, msg: String?, isHtml: Boolean) {
        Handler(Looper.getMainLooper()).post {
            if (isHtml) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    Toast.makeText(
                        context,
                        Html.fromHtml(msg, Html.FROM_HTML_MODE_LEGACY),
                        Toast.LENGTH_LONG
                    ).show()
                }
            } else {
                Toast.makeText(context, msg, Toast.LENGTH_LONG).show()
            }
        }
    }

    /**
     * toast with layout
     */
    @JvmStatic
    fun setToastLayout(activity: Activity, msg: String?, context: Context?) {
        val inflater = activity.layoutInflater
        val layout =
            inflater.inflate(R.layout.toast_layout, activity.findViewById(R.id.toastLayout))

        val toast = Toast.makeText(context, msg, Toast.LENGTH_SHORT)
        toast.setGravity(Gravity.BOTTOM, 0, 100)
        toast.duration = Toast.LENGTH_SHORT
        toast.view = layout

        toast.show()
    }
}